design/open/delta-diff.md

```diff
@@ -0,0 +1,160 @@
+
+![Microkernel architecture overview](diagrams/microkernel-overview.png)
+[(excalidraw file)](diagrams/microkernel-overview.excalidraw)
+
+---
+
+## Implementation
+
+![Delta Diff flow](diagrams/delta-diff-flow.png)
+[(excalidraw file)](diagrams/delta-diff-flow.excalidraw)
+
+#### DiffService
+
+The DiffService will be an internal component in lakeFS which will serve as the Core system.  
+In order to realize which diff plugins are available, we shall use new configuration values:
+1. `LAKEFS_PLUGINS_LOCATION`
+   The DiffService will use this location to search for the required plugin. Defaults to lakeFS's binary directory.
+2. `LAKEFS_PLUGIN_DIFF_{TYPE OF DIFF}`
+   The DiffService will use the name of binary provided by this env var to perform the diff. For instance, `LAKEFS_PLUGIN_DIFF_DELTA=deltaDiffBinary`
+   The DiffService will use the [plugins' location](https://github.com/hashicorp/terraform/blob/main/plugins.go) and the diff binary to load the plugin and request a diff.
+   The **type** of diff will be sent as part of the request to lakeFS as specified [here](#API).
+   The communication between the DiffService and the plugins, as explained above, will be through the `go-plugin` package (`gRPC`).
+
+#### Delta Diff Plugin
+
+Implemented using [delta-rs](https://github.com/delta-io/delta-rs) (Rust), this plugin will perform the diff operation using table paths provided by the DiffService through a `gRPC` call.  
+To query the Delta Table from lakeFS, the plugin will generate an S3 client (this is a constraint imposed by the `delta-rs` package) and send a request to lakeFS's S3 gateway.  
+The diff algorithm:
+1. Run the Delta [HISTORY command](https://docs.delta.io/latest/delta-utility.html#history-schema) on both table paths.
+2. Traverse through the [returned "commitInfo" entry vector ](https://github.com/delta-io/delta-rs/blob/main/rust/src/delta.rs#L888)
+    1. While the returned entry versions are equal:
+        1. Create a hash for the entry based on fields: `timestamp`, `operation`, `operationParameters` , and `operationMetrics` values.
+        2. Compare the hashes of the versions.
+        3. If they're equal, continue, else stop.
+3. Return the "left"'s path history vector from the stopped location.
+
+### Authentication
+
+The `delta-rs` package generates an S3 client which will be used to communicate back to lakeFS (through the S3 gateway).  
+In order for the S3 client to communicate with lakeFS (or S3 in general) it needs to pass an AWS Access Key Id and Secret access key.  
+Since applicative credentials are not obligatory, the users that sent the request for a diff might not have such, and even if they have,
+they cannot send them through the GUI (which is the UI we chose to implement this feature for the MVP).
+To overcome this scenario, we'll use special diff credentials as follows:
+1. User makes a Delta Diff request.
+2. The DiffService checks if the user has "diff credentials" in the DB:
```

Comment:
This is essentially a session. Maybe we should implement a session mechanism as part of this feature?